FishTracker/tracker.py

```diff
@@ -0,0 +1,348 @@
+"""
+Functions to track wave-type electric fish frequencies over longer periods of time.
+
+fish_tracker(): main function which performs all steps including loading data, fish tracking and -sorting and more.
+"""
+import sys
+import numpy as np
+import matplotlib.pyplot as plt
+import thunderfish.dataloader as dl
+import thunderfish.powerspectrum as ps
+import thunderfish.harmonicgroups as hg
+import thunderfish.config_tools as ct
+
+
+def first_level_fish_sorting(all_fundamentals, audio_file, all_times, max_time_tolerance=5., freq_tolerance = 2., save_original_fishes=False):
+    """
+    Sorts fundamental frequencies of wave-type electric fish detected at certain timestamps to fishes.
+
+    There is an array of fundamental frequencies for every timestamp (all_fundamentals). Each of these frequencies is
+    compared to the last frequency of already detected fishes (last_fish_fundamentals). If the frequency difference
+    between the new frequency and one or multiple already detected fishes the frequency is appended to the array
+    containing all frequencies of the fish (fishes) that has been absent for the shortest period of time. If the
+    frequency doesn't fit to one fish, a new fish array is created. If a fish has not been detected at one time-step
+    a NaN is added to this fish array.
+
+    The result is for each fish a array containing frequencies or nans with the same length than the time array
+    (all_times). These fish arrays can be saved as .npy file to access the code after the time demanding step.
+
+    :param all_fundamentals: (list) containing arrays with the fundamentals frequencies of fishes detected at a certain time.
+    :param audio_file: (string) filepath of the analysed audiofile.
+    :param all_times: (array) containing time stamps of frequency detection. (  len(all_times) == len(fishes[xy])  )
+    :param max_time_tolerance: (int) time in minutes from when a certain fish is no longer tracked.
+    :param freq_tolerance: (float) maximum frequency difference to assign a frequency to a certain fish.
+    :param save_original_fishes: (boolean) if True saves the sorted fishes after the first level of fish sorting.
+    :return fishes: (list) containing arrays of sorted fish frequencies. Each array represents one fish.
+    """
+    # ToDo: insert clean up function that deletes fishes with less than 5 detections after every hour...
+    detection_time_diff = np.median(np.diff(all_times))
+    dpm = 60. / detection_time_diff  # detections per minutes
+
+    fishes = [[ 0. ]]
+    last_fish_fundamentals = [ 0. ]
+    end_nans = [0]
+
+    # for every list of fundamentals ...
+    print len(all_fundamentals)
+    for t_list in range(len(all_fundamentals)):
```

Comment:
no. now that i have a clean up function it is also used as an independent index.